sql/hive-thriftserver/src/main/java/org/apache/hive/service/auth/HiveAuthFactory.java

```diff
@@ -362,4 +371,34 @@ public static void verifyProxyAccess(String realUser, String proxyUser, String i
     }
   }
 
+  public static boolean needUgiLogin(UserGroupInformation ugi, String principal, String keytab) {
+    return null == ugi || !ugi.hasKerberosCredentials() || !ugi.getUserName().equals(principal) ||
+      !keytab.equals(getKeytabFromUgi());
+  }
+
+  private static String getKeytabFromUgi() {
+    Class<?> clz = UserGroupInformation.class;
+    try {
+      synchronized (clz) {
+        Field field = clz.getDeclaredField("keytabFile");
```

Comment:
The field lookup + setAccessible can be pulled out of the method and into a static variable.
The field.get() itself should be within the sync block.